File: examples/cs/CustomThrottler.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using ccxt;

namespace CustomThrottlerExamples
{
    // Example 1: Simple delay-based throttler
    public class SimpleDelayThrottler : ICustomThrottler
    {
        private readonly int delayMs;

        public SimpleDelayThrottler(int delayMs = 1000)
        {
            this.delayMs = delayMs;
        }

        public async Task Throttle(object cost = null)
        {
            // Simple delay regardless of cost
            await Task.Delay(delayMs);
        }
    }

    // Example 2: Token bucket throttler with custom logic
    public class CustomTokenBucketThrottler : ICustomThrottler
    {
        private double tokens;
        private readonly double capacity;
        private readonly double refillRate;
        private DateTime lastRefill;
        private readonly object lockObject = new object();

        public CustomTokenBucketThrottler(double capacity = 10.0, double refillRate = 1.0)
        {
            this.tokens = capacity;
            this.capacity = capacity;
            this.refillRate = refillRate;
            this.lastRefill = DateTime.UtcNow;
        }

        public async Task Throttle(object cost = null)
        {
            double requestCost = cost != null ? Convert.ToDouble(cost) : 1.0;

            lock (lockObject)
            {
                // Refill tokens
                var now = DateTime.UtcNow;
                var elapsed = (now - lastRefill).TotalSeconds;
                var tokensToAdd = elapsed * refillRate;
                tokens = Math.Min(capacity, tokens + tokensToAdd);
                lastRefill = now;

                // Check if we have enough tokens
                if (tokens < requestCost)
                {
                    // Calculate wait time
                    var waitTime = (requestCost - tokens) / refillRate;
                    tokens = 0;
                    
                    // Wait outside the lock
                    Task.Delay(TimeSpan.FromSeconds(waitTime)).Wait();
                }
                else
                {
                    tokens -= requestCost;
                }
            }
        }
    }

    // Example 3: Adaptive throttler that adjusts based on response times
    public class AdaptiveThrottler : ICustomThrottler
    {
        private TimeSpan baseDelay;
        private TimeSpan maxDelay;
        private TimeSpan currentDelay;
        private int successCount;
        private int errorCount;
        private readonly object lockObject = new object();

        public AdaptiveThrottler(TimeSpan baseDelay, TimeSpan maxDelay)
        {
            this.baseDelay = baseDelay;
            this.maxDelay = maxDelay;
            this.currentDelay = baseDelay;
        }

        public async Task Throttle(object cost = null)
        {
            TimeSpan delay;
            lock (lockObject)
            {
                delay = currentDelay;
            }

            await Task.Delay(delay);
        }

        public void OnSuccess()
        {
            lock (lockObject)
            {
                successCount++;
                if (successCount >= 5)
                {
                    // Reduce delay on success
                    currentDelay = TimeSpan.FromMilliseconds(Math.Max(baseDelay.TotalMilliseconds, currentDelay.TotalMilliseconds / 2));
                    successCount = 0;
                    errorCount = 0;
                }
            }
        }

        public void OnError()
        {
            lock (lockObject)
            {
                errorCount++;
                if (errorCount >= 3)
                {
                    // Increase delay on errors
                    currentDelay = TimeSpan.FromMilliseconds(Math.Min(maxDelay.TotalMilliseconds, currentDelay.TotalMilliseconds * 2));
                    successCount = 0;
                    errorCount = 0;
                }
            }
        }
    }

    // Example 4: Exchange-aware throttler
    public class ExchangeAwareThrottler : ICustomThrottler
    {
        private readonly Dictionary<string, TimeSpan> exchangeRates;
        private readonly TimeSpan defaultRate;
        private readonly object lockObject = new object();

        public ExchangeAwareThrottler()
        {
            exchangeRates = new Dictionary<string, TimeSpan>
            {
                { "binance", TimeSpan.FromMilliseconds(100) },
                { "coinbase", TimeSpan.FromMilliseconds(500) },
                { "kraken", TimeSpan.FromMilliseconds(200) }
            };
            defaultRate = TimeSpan.FromMilliseconds(1000);
        }

        public async Task Throttle(object cost = null)
        {
            // Use default rate (in real implementation, you'd get exchange name from context)
            await Task.Delay(defaultRate);
        }

        public TimeSpan GetRateForExchange(string exchangeId)
        {
            lock (lockObject)
            {
                return exchangeRates.TryGetValue(exchangeId, out var rate) ? rate : defaultRate;
            }
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("=== CCXT C# Custom Throttler Examples ===\n");

            // Example 1: Simple delay throttler
            Console.WriteLine("1. Simple Delay Throttler Example:");
            var simpleThrottler = new SimpleDelayThrottler(500); // 500ms delay
            
            // Create exchange with custom throttler
            var exchange = new Binance(new Dictionary<string, object>
            {
                { "customThrottler", simpleThrottler },
                { "enableRateLimit", true }
            });

            // Test the throttler
            var start = DateTime.UtcNow;
            await simpleThrottler.Throttle(1.0);
            var elapsed = DateTime.UtcNow - start;
            Console.WriteLine($"   Throttled request took: {elapsed.TotalMilliseconds:F0}ms");

            // Example 2: Token bucket throttler
            Console.WriteLine("\n2. Token Bucket Throttler Example:");
            var tokenThrottler = new CustomTokenBucketThrottler(5.0, 1.0); // 5 tokens, 1 token/sec refill
            
            start = DateTime.UtcNow;
            for (int i = 0; i < 3; i++)
            {
                await tokenThrottler.Throttle(2.0); // Each request costs 2 tokens
                Console.WriteLine($"   Request {i + 1} completed");
            }
            elapsed = DateTime.UtcNow - start;
            Console.WriteLine($"   Total time for 3 requests: {elapsed.TotalMilliseconds:F0}ms");

            // Example 3: Adaptive throttler
            Console.WriteLine("\n3. Adaptive Throttler Example:");
            var adaptiveThrottler = new AdaptiveThrottler(TimeSpan.FromMilliseconds(100), TimeSpan.FromSeconds(2));
            
            start = DateTime.UtcNow;
            for (int i = 0; i < 3; i++)
            {
                await adaptiveThrottler.Throttle(1.0);
                adaptiveThrottler.OnSuccess(); // Simulate successful request
                Console.WriteLine($"   Adaptive request {i + 1} completed");
            }
            elapsed = DateTime.UtcNow - start;
            Console.WriteLine($"   Total time for 3 adaptive requests: {elapsed.TotalMilliseconds:F0}ms");

            // Example 4: Exchange-aware throttler
            Console.WriteLine("\n4. Exchange-Aware Throttler Example:");
            var exchangeThrottler = new ExchangeAwareThrottler();
            
            start = DateTime.UtcNow;
            await exchangeThrottler.Throttle(1.0);
            elapsed = DateTime.UtcNow - start;
            Console.WriteLine($"   Exchange-aware request took: {elapsed.TotalMilliseconds:F0}ms");

            Console.WriteLine("\n=== Examples completed ===");
        }
    }
}

File: examples/go/custom-throttler.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/ccxt/ccxt/go/v4"
)

// Example 1: Simple delay-based throttler
type SimpleDelayThrottler struct {
	delayMs int64
}

func NewSimpleDelayThrottler(delayMs int64) *SimpleDelayThrottler {
	return &SimpleDelayThrottler{delayMs: delayMs}
}

func (t *SimpleDelayThrottler) Throttle(ctx context.Context, cost float64) error {
	time.Sleep(time.Duration(t.delayMs) * time.Millisecond)
	return nil
}

// Example 2: Token bucket throttler with custom logic
type CustomTokenBucketThrottler struct {
	tokens     float64
	capacity   float64
	refillRate float64
	lastRefill time.Time
	mutex      chan struct{}
}

func NewCustomTokenBucketThrottler(capacity, refillRate float64) *CustomTokenBucketThrottler {
	return &CustomTokenBucketThrottler{
		tokens:     capacity,
		capacity:   capacity,
		refillRate: refillRate,
		lastRefill: time.Now(),
		mutex:      make(chan struct{}, 1),
	}
}

func (t *CustomTokenBucketThrottler) Throttle(ctx context.Context, cost float64) error {
	t.mutex <- struct{}{} // Lock
	defer func() { <-t.mutex }() // Unlock

	// Refill tokens
	now := time.Now()
	elapsed := now.Sub(t.lastRefill).Seconds()
	tokensToAdd := elapsed * t.refillRate
	t.tokens = min(t.capacity, t.tokens+tokensToAdd)
	t.lastRefill = now

	// Check if we have enough tokens
	if t.tokens < cost {
		// Calculate wait time
		waitTime := (cost - t.tokens) / t.refillRate
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(time.Duration(waitTime * float64(time.Second))):
		}
		t.tokens = 0
	} else {
		t.tokens -= cost
	}

	return nil
}

// Example 3: Adaptive throttler that adjusts based on response times
type AdaptiveThrottler struct {
	baseDelay    time.Duration
	maxDelay     time.Duration
	currentDelay time.Duration
	successCount int
	errorCount   int
	mutex        chan struct{}
}

func NewAdaptiveThrottler(baseDelay, maxDelay time.Duration) *AdaptiveThrottler {
	return &AdaptiveThrottler{
		baseDelay:    baseDelay,
		maxDelay:     maxDelay,
		currentDelay: baseDelay,
		mutex:        make(chan struct{}, 1),
	}
}

func (t *AdaptiveThrottler) Throttle(ctx context.Context, cost float64) error {
	t.mutex <- struct{}{} // Lock
	defer func() { <-t.mutex }() // Unlock

	// Use current delay
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(t.currentDelay):
	}

	return nil
}

func (t *AdaptiveThrottler) OnSuccess() {
	t.mutex <- struct{}{} // Lock
	defer func() { <-t.mutex }() // Unlock

	t.successCount++
	if t.successCount >= 5 {
		// Reduce delay on success
		t.currentDelay = max(t.baseDelay, t.currentDelay/2)
		t.successCount = 0
		t.errorCount = 0
	}
}

func (t *AdaptiveThrottler) OnError() {
	t.mutex <- struct{}{} // Lock
	defer func() { <-t.mutex }() // Unlock

	t.errorCount++
	if t.errorCount >= 3 {
		// Increase delay on errors
		t.currentDelay = min(t.maxDelay, t.currentDelay*2)
		t.successCount = 0
		t.errorCount = 0
	}
}

// Example 4: Exchange-aware throttler
type ExchangeAwareThrottler struct {
	exchangeRates map[string]time.Duration
	defaultRate   time.Duration
	mutex         chan struct{}
}

func NewExchangeAwareThrottler() *ExchangeAwareThrottler {
	return &ExchangeAwareThrottler{
		exchangeRates: map[string]time.Duration{
			"binance": 100 * time.Millisecond,
			"coinbase": 500 * time.Millisecond,
			"kraken":  200 * time.Millisecond,
		},
		defaultRate: 1000 * time.Millisecond,
		mutex:       make(chan struct{}, 1),
	}
}

func (t *ExchangeAwareThrottler) Throttle(ctx context.Context, cost float64) error {
	t.mutex <- struct{}{} // Lock
	defer func() { <-t.mutex }() // Unlock

	// Use default rate (in real implementation, you'd get exchange name from context)
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(t.defaultRate):
	}

	return nil
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func max(a, b time.Duration) time.Duration {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println("=== CCXT Go Custom Throttler Examples ===\n")

	// Example 1: Simple delay throttler
	fmt.Println("1. Simple Delay Throttler Example:")
	simpleThrottler := NewSimpleDelayThrottler(500) // 500ms delay
	
	// Create exchange with custom throttler
	exchange := ccxt.NewBinance(map[string]interface{}{
		"customThrottler": simpleThrottler,
		"enableRateLimit": true,
	})

	// Test the throttler
	start := time.Now()
	ctx := context.Background()
	
	err := simpleThrottler.Throttle(ctx, 1.0)
	if err != nil {
		log.Printf("Error: %v", err)
	}
	
	elapsed := time.Since(start)
	fmt.Printf("   Throttled request took: %v\n", elapsed)

	// Example 2: Token bucket throttler
	fmt.Println("\n2. Token Bucket Throttler Example:")
	tokenThrottler := NewCustomTokenBucketThrottler(5.0, 1.0) // 5 tokens, 1 token/sec refill
	
	start = time.Now()
	for i := 0; i < 3; i++ {
		err := tokenThrottler.Throttle(ctx, 2.0) // Each request costs 2 tokens
		if err != nil {
			log.Printf("Error: %v", err)
		}
		fmt.Printf("   Request %d completed\n", i+1)
	}
	elapsed = time.Since(start)
	fmt.Printf("   Total time for 3 requests: %v\n", elapsed)

	// Example 3: Adaptive throttler
	fmt.Println("\n3. Adaptive Throttler Example:")
	adaptiveThrottler := NewAdaptiveThrottler(100*time.Millisecond, 2*time.Second)
	
	start = time.Now()
	for i := 0; i < 3; i++ {
		err := adaptiveThrottler.Throttle(ctx, 1.0)
		if err != nil {
			log.Printf("Error: %v", err)
		}
		adaptiveThrottler.OnSuccess() // Simulate successful request
		fmt.Printf("   Adaptive request %d completed\n", i+1)
	}
	elapsed = time.Since(start)
	fmt.Printf("   Total time for 3 adaptive requests: %v\n", elapsed)

	// Example 4: Exchange-aware throttler
	fmt.Println("\n4. Exchange-Aware Throttler Example:")
	exchangeThrottler := NewExchangeAwareThrottler()
	
	start = time.Now()
	err = exchangeThrottler.Throttle(ctx, 1.0)
	if err != nil {
		log.Printf("Error: %v", err)
	}
	elapsed = time.Since(start)
	fmt.Printf("   Exchange-aware request took: %v\n", elapsed)

	fmt.Println("\n=== Examples completed ===")
}